### PR TITLE
Fix for #Pydev 646: Don't complain if self is not in parameters for zope.interface.Interface

### PR DESCRIPTION
**Pull request by oluiscabral**

### CODE REVIEW

File: plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/visitors/NoSelfChecker.java [hunk 1]
```diff
@@ -176,6 +177,22 @@ public void beforeFunctionDef(FunctionDef node) {
         scope.push(Scope.SCOPE_TYPE_METHOD);
     }
 
+    private static boolean hasInterfaceInBase(FunctionDef node) {
```
Comment (wesleybl) [inline plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/visitors/NoSelfChecker.java:180]: Will this method work for classes that don't directly inherit from `zope.interface.Interface`? For example:

```python
from zope.interface import Interface

class Ia(Interface):
    def method_a():  # Shouldn't complain about the lack of self
        """"""

class Ib(Ia):
    def method_b():  # Shouldn't complain about the lack of self
        """"""

```

File: plugins/org.python.pydev/tests_analysis/com/python/pydev/analysis/OccurrencesAnalyzer2Test.java [hunk 2]
```diff
@@ -846,4 +846,27 @@ public void testInvalidEncoding() throws Exception {
...
+                + "        print \"->\", self, \"Object\"\n"
+                + "c = Ic()\n"
+                + "print type(I), type(Ic), type(c), type(Ic.F), type(c.F)\n"
+                + "c.F()");
+        checkError("Undefined variable: Interface", "Undefined variable: implementer");
+    }
+
+    public void testNoSelfInSomeInterface() throws Exception {
+        doc = new Document(""
+                + "class I('Interface'):\n"
```
Comment (fabioz) [inline plugins/org.python.pydev/tests_analysis/com/python/pydev/analysis/OccurrencesAnalyzer2Test.java:867]: This test doesn't seem correct... I don't think you can really inherit from a string...

File: plugins/org.python.pydev/tests_analysis/com/python/pydev/analysis/OccurrencesAnalyzer2Test.java [hunk 3]
```diff
@@ -846,4 +846,27 @@ public void testInvalidEncoding() throws Exception {
     //        );
     //        checkError(1);
     //    }
+
+    public void testNoSelfInZopeInterface() throws Exception {
+        doc = new Document(""
+                + "class I(Interface):\n"
```
Comment (fabioz) [inline plugins/org.python.pydev/tests_analysis/com/python/pydev/analysis/OccurrencesAnalyzer2Test.java:852]: I think that you should shouldn't really do the test without the actual import:

```
from zope.interface import Interface
```

the problem is that there may be other libraries that have `Interface` declared and you don't want to skip the `self` check on those... something such as:

```
from zope.interface import Interface as I

class Impl(I):
    def method_a():
        pass
```

must also work.

Please make a separate test for the case above too...

File: plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/visitors/NoSelfChecker.java [hunk 4]
```diff
@@ -176,6 +178,28 @@ public void beforeFunctionDef(FunctionDef node) {
...
+        try {
+            ClassDef parent = (ClassDef) node.parent;
+            for (exprType base : parent.bases) {
+                String rep = null;
+                if (base instanceof Str) {
```
Comment (fabioz) [inline plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/visitors/NoSelfChecker.java:186]: This doesn't seem correct (I don't think you can inherit from a string in Python).

File: plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/visitors/NoSelfChecker.java (hunk 1 above)
Comment (fabioz) [reply plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/visitors/NoSelfChecker.java:181]: This doesn't work right now, but it should (so, we should have a test-case for this use-case).

File: plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/visitors/NoSelfChecker.java [hunk 5]
```diff
@@ -159,7 +160,8 @@ public void beforeFunctionDef(FunctionDef node) {
                             new Tuple<Expected, FunctionDef>(new Expected("self or cls", received), node));
                 }
 
-            } else if (!startsWithSelf && !startsWithCls && !isStaticMethod && !isClassMethod) {
+            } else if (!startsWithSelf && !startsWithCls && !isStaticMethod && !isClassMethod
+                    && !hasInterfaceInBase(node)) {
```
Comment (fabioz) [inline plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/visitors/NoSelfChecker.java:164]: You shouldn't be checking this for each method, rather, compute the `hasInterfaceInBase` only once (it should probably be `hasZopeInterfaceInBase` as other libraries may also be `Interface`).

Also, it shouldn't just check for the string, rather it should find the definition... in `com.python.pydev.analysis.scopeanalysis.AbstractScopeAnalyzerVisitor.visitClassDef(ClassDef)`, when visiting the bases (when it's already finding the definition of the names) whether a given name will end up mapping to a zope interface and then store that info (as it's a visitor, you should push/pop that it's visiting bases and then on `markRead` if it's visiting bases mark whether it'll end up mapping to a zope interface if that's the case -- that should also be stored in a stack and popped appropriately when the class visit finishes).

File: plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/visitors/NoSelfChecker.java [hunk 6]
```diff
@@ -176,6 +178,28 @@ public void beforeFunctionDef(FunctionDef node) {
...
+    private static boolean hasInterfaceInBase(FunctionDef node) {
+        try {
+            ClassDef parent = (ClassDef) node.parent;
+            for (exprType base : parent.bases) {
+                String rep = null;
+                if (base instanceof Str) {
+                    Str str = (Str) base;
+                    rep = str.s;
+                } else {
+                    rep = NodeUtils.getFullRepresentationString(base);
```
Comment (fabioz) [inline plugins/com.python.pydev.analysis/src/com/python/pydev/analysis/visitors/NoSelfChecker.java:190]: This isn't the ideal approach, we should actually follow the import at the class definition (see the other comment related to this).

**Review by fabioz — CHANGES_REQUESTED** [review]

**Comment by fabioz** [comment]
This was integrated in: https://github.com/fabioz/Pydev/commit/2b768f67348bf2ca43b91993afcb3ade09824581

Note that there were still extensive changes while I was reviewing, so, I ended up doing a single commit instead of doing a commit and then fixing it in a subsequent commit.

**Pull request closed without merging**